zephyr/esp32s3/src/bt/esp_bt_adapter.c

```diff
@@ -0,0 +1,1434 @@
+	if (uncertainty < BTDM_MIN_TIMER_UNCERTAINTY_US) {
+		uncertainty = BTDM_MIN_TIMER_UNCERTAINTY_US;
+	}
+
+	assert (s_lp_stat.wakeup_timer_started == 0);
+	if (esp_timer_start_once(s_btdm_slp_tmr, us_to_sleep - uncertainty) == ESP_OK) {
+		s_lp_stat.wakeup_timer_started = 1;
+	} else {
+		LOG_ERR("timer start failed");
+		assert(0);
+	}
+}
+
+static void btdm_sleep_enter_phase2_wrapper(void)
+{
+	if (btdm_controller_get_sleep_mode() == ESP_BT_SLEEP_MODE_1) {
+		if (s_lp_stat.phy_enabled) {
+			esp_phy_disable();
+			s_lp_stat.phy_enabled = 0;
+		} else {
+			assert(0);
+		}
+
+		if (s_lp_stat.pm_lock_released == 0) {
+			s_lp_stat.pm_lock_released = 1;
+		}
+	}
+}
+
+static void btdm_sleep_exit_phase3_wrapper(void)
+{
+	if (btdm_controller_get_sleep_mode() == ESP_BT_SLEEP_MODE_1) {
+		if (s_lp_stat.phy_enabled == 0) {
+			esp_phy_enable();
+			s_lp_stat.phy_enabled = 1;
+		}
+	}
+
+	// If BT wakeup before esp timer coming due to timer task have no chance to run.
+	// Then we will not run into `btdm_sleep_exit_phase0` and stop esp timer,
+	// Do it again here to fix this issue.
+	if (s_lp_cntl.wakeup_timer_required && s_lp_stat.wakeup_timer_started) {
+		esp_timer_stop(s_btdm_slp_tmr);
+		s_lp_stat.wakeup_timer_started = 0;
+	}
+
+	// wait for the sleep state to change
+	// the procedure duration is at micro-second level or less
+	while (btdm_sleep_clock_sync()) {
+		;
+	}
+}
+
+static void IRAM_ATTR btdm_sleep_exit_phase0(void *param)
+{
+	assert(s_lp_cntl.enable == 1);
+
+	int event = (int) param;
+	if (event == BTDM_ASYNC_WAKEUP_SRC_VHCI || event == BTDM_ASYNC_WAKEUP_SRC_DISA) {
+		btdm_wakeup_request();
+	}
+
+	if (s_lp_cntl.wakeup_timer_required && s_lp_stat.wakeup_timer_started) {
+		esp_timer_stop(s_btdm_slp_tmr);
+		s_lp_stat.wakeup_timer_started = 0;
+	}
+
+	if (event == BTDM_ASYNC_WAKEUP_SRC_VHCI || event == BTDM_ASYNC_WAKEUP_SRC_DISA) {
+		semphr_give_wrapper(s_wakeup_req_sem);
+	}
+}
+
+static void IRAM_ATTR btdm_slp_tmr_callback(void *arg)
+{
+#ifdef CONFIG_PM_ENABLE
+	btdm_vnd_offload_post(BTDM_VND_OL_SIG_WAKEUP_TMR, (void *)BTDM_ASYNC_WAKEUP_SRC_TMR);
+#endif
+}
+
+static bool async_wakeup_request(int event)
+{
+	if (s_lp_cntl.enable == 0) {
+		return false;
+	}
+
+	bool do_wakeup_request = false;
+	switch (event) {
+		case BTDM_ASYNC_WAKEUP_SRC_VHCI:
+		case BTDM_ASYNC_WAKEUP_SRC_DISA:
+			btdm_in_wakeup_requesting_set(true);
+			if (!btdm_power_state_active()) {
+				btdm_vnd_offload_post(BTDM_VND_OL_SIG_WAKEUP_TMR, (void *)event);
+				do_wakeup_request = true;
+				semphr_take_wrapper(s_wakeup_req_sem, OSI_FUNCS_TIME_BLOCKING);
+			}
+			break;
+		default:
+			break;
+	}
+
+	return do_wakeup_request;
+}
+
+static void async_wakeup_request_end(int event)
+{
+	if (s_lp_cntl.enable == 0) {
+		return;
+	}
+
+	bool allow_to_sleep;
+	switch (event) {
+		case BTDM_ASYNC_WAKEUP_SRC_VHCI:
+		case BTDM_ASYNC_WAKEUP_SRC_DISA:
+			allow_to_sleep = true;
+			break;
+		default:
+			allow_to_sleep = true;
```

Comment:
Shouldn't it be allow_to_sleep = false on default?

Reply:
I know it seems odd. This is based on original file and I suggest not changing this for now:  
https://github.com/espressif/esp-idf/blob/release/v4.4/components/bt/controller/esp32s3/bt.c#L856